File: Python/340__Longest_Substring_with_At_Most_K_Distinct_Characters.py
class Solution(object):
    def lengthOfLongestSubstringKDistinct(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        if s == "":
            return 0
        if k >= len(s):
            return len(s)
        if k == 0:
            return 0
        start, end = 0, 0
        result = 0
        dic = {}
        dic[s[0]] = 1

        while end < len(s):

            while len(dic) <= k and end < len(s) - 1:
                if len(dic) != k:
                    end += 1
                    if s[end] in dic:
                        dic[s[end]] += 1
                    else:
                        dic[s[end]] = 1
                else:
                    if s[end + 1] in dic:
                        end += 1
                        dic[s[end]] += 1
                    else:
                        break;
            result = max(result, end - start + 1)
            if end == len(s) - 1:
                break
            while len(dic) >= k:
                if dic[s[start]] == 1:
                    del dic[s[start]]
                else:
                    dic[s[start]] -= 1
                start += 1
        return result


# The solution above can be optimized to :

class Solution(object):
    def lengthOfLongestSubstringKDistinct(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        d = {}
        low, ret = 0, 0
        for i, c in enumerate(s):
            d[c] = i
            if len(d) > k:
                low = min(d.values())
                del d[s[low]]
                low += 1
            ret = max(i - low + 1, ret)
        return ret

File: Java/340__Longest_Substring_with_At_Most_K_Distinct_Characters.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        if (s.length() == 0 || k == 0) {
            return 0;
        }
        HashMap<Character, Integer> map = new HashMap<>();
        int result = 1;
        int l = 0, r = 0;
        while (r < s.length()) {
            char cur = s.charAt(r);
            map.put(cur, r++);
            if (map.size() > k) {
                int delete_i = Collections.min(map.values());
                map.remove(s.charAt(delete_i));
                l = delete_i + 1;
            }
            result = Math.max(result, r - l);
        }
        return result;
    }
}
